### ISSUE TITLE
Q about installer and updater: the files are NOT signed?

### ISSUE THREAD

**Issue by COLABORATI** [netdata/netdata#3793]
From what I understand the [binary] installer and updater do not check for any file signatures - so it looks like this might be a great entry to anybodies infrastructure, as it might be assumed that netdata will be installed on many machines this is a very attractive target.

I hope I am wrong? If not, wouldn´t it be great to provide signed packages for the ost used operating systems at least to give people a better way to install this software in a secure manner?

**Comment by Ferroin** [comment]
For the install and upgrade scripts, you're correct.

For systems which already have official packages (that is, those 'most used operating systems' you mention), the distribution's existing infrastructure either already does this (this is the case for most of them) or is in the process of being updated to do this.

**Comment by COLABORATI** [comment]
OMG, please at least put a big red warning on the installer - of course experienced admins will recognize that but I guess thousands of noobs are installing that software right from the copy paste  script and their machines are now easy targets for some takeover because *every day* the update script generates a chance to implant some unsigned code. Very bad idea. You should know better.

It would be better to remove that install procedure completely - you are acting very irresponsible here - how could you even do that? Also this is a really bad indicator for the 'general security awareness' factor for this project - I would not have expected that by the author. 

Opensuse has a good build service, also Launchpad helps a lot with creating PPAs, so there should be no reason to go on with that bad practice. I mean, common, the idea of 'signing packages' is not really new.

**Comment by Ferroin** [comment]
Netdata is already packaged in almost all major distributions (though the wiki page on installation needs updated to reflect this).  The scripts you're complaining about are a last-resort fallback for people who either:

* Are running a distribution which doesn't have it packaged and can't use a package from another distribution.
* Are running a distribution that doesn't have a package manager.
* Are running a distribution which provides no way to build netdata locally (in the case of `kickstart-static64.sh`.

Note that I'm not saying those shouldn't have verification, just that it's _really_ not a priority given that most people won't be using them.

**Comment by ktsaou** [comment]
hm... how do you suggest this verification to happen?

I mean, the scripts are downloaded over SSL and all web clients verify the identity of the transmitting web server - so you know they are downloaded from our servers.

Then you suggest we should also verify the signature. ok, I like that.
But how?

The signature will be downloaded from the same SSL web server?
This does not seem much of a protection. If an attacker manages to fake our web server (including SSL verification), I don't see why he will not be able to inject his own signature.

I am open to other ideas...

**Comment by ktsaou** [comment]
btw, the updater script is generated from the installer from the previous run. So, it is a dynamically generated script, custom for each installation.

**Comment by Ferroin** [comment]
Strictly speaking, checking the hash still improves security by a marginal amount.  Statistically, it's harder to perform two MitM attacks in a row than it is to perform just one, so there's a possibility that one or the other will not be compromised.  However, this depends on the client not pipelining the requests.

**Comment by ktsaou** [comment]
hm...we could download the signature from github directly.
But then we have this problem:

1. The user runs `kickstart.sh` downloaded from https://my-netdata.io and runs it.
2. How the signature of this script will be checked? I guess only by itself. So, the downloaded script will download the signature from github and check itself (???).

hm... not much of a protection... since the downloaded script will verify itself. If the attacker manages to overcome SSL verification and injects his own script, it is obvious that his script will not verify itself...

**Comment by ktsaou** [comment]
Guys, I don't think there is a solution to this.

Keep in mind this "unsecure" way is used only once, at the **initial installation**.

All updates are through git, and these updates have a ton of signatures. So, if you install the legit repo at the beginning, all updates are checked against the repo signatures, commit by commit.

I don't see how we can improve this... of course I am open to alternatives if you believe I am missing something.

**Comment by csamsel** [comment]
Using an curl pipe to bash command is always super insecure. Signing some files only helps marginally.

Use your distributions package manager to install netdata.

**Comment by paulfantom** [comment]
We are working on providing binary packages. Tracking current progress on this matter can be done by visiting this github project: https://github.com/netdata/netdata/projects/13.

Binary packages will be signed, but until then there is no other method of installing netdata apart from shell script or docker container.
